src/Filters/WhereSearch.php

```diff
@@ -0,0 +1,81 @@
+<?php
+/*
+ * Copyright 2022 Cloud Creativity Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+declare(strict_types=1);
+
+namespace LaravelJsonApi\Eloquent\Filters;
+
+use LaravelJsonApi\Eloquent\Contracts\Filter;
+use LaravelJsonApi\Eloquent\Filters\Concerns\DeserializesValue;
+
+class WhereSearch implements Filter
+{
+    use DeserializesValue;
+    use Concerns\IsSingular;
+
+    /**
+     * @var string
+     */
+    private string $name;
+
+    /**
+     * Create a new filter.
+     *
+     * @param string $name
+     * @param string|null $column
+     * @return static
+     */
+    public static function make(string $name, string $columns): self
+    {
+        return new static($name, $columns);
+    }
+
+    /**
+     * WhereSearch constructor.
+     *
+     * @param string $name
+     * @param string $column
+     */
+    public function __construct(string $name, string $columns)
+    {
+        $this->name = $name;
+        $this->column = $columns;
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function key(): string
+    {
+        return $this->name;
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function apply($query, $value)
+    {
+        $searchableColumns = explode('|', $this->column);
+        $value = $this->deserialize($value);
+
+        foreach ($searchableColumns as $searchableColumn) {
```

Comment:
We need different logic here based on there being 1 or many columns. We also need to qualify the column name.

One is simple:

```php
if (1 === count($this->columns)) {
    $query->where(
        $query->qualifyColumn($this->columns[0]),
        'LIKE',
        $this->pattern($value)
    );
    return $query;
}
```

For multiple, we can't use `orWhere` directly on the query, because other filters might be adding where clauses. So we need the all the `orWhere` to be wrapped in a brackets for the database query. Something like:

```php
$pattern = $this->pattern($value);

$query->where(function ($q) use ($pattern) {
    foreach ($this->columns as $column) {
        $q->orWhere($q->qualifyColumn($column), 'LIKE', $pattern);
    }
});
```

I think that code works but I'm typing directly into Github so you'll have to check my suggestions actually work.

For the `pattern()` method, that should take the value and convert it to the pattern we'll use to query the database. We'll default to putting the `%` at the start and end of the value, as you've done for this. But we should let the developer specify whether they only want to search for the value at the start or the end of the column value.

I.e. add a `private ?string $position` property. Then:

```php
public function beginsWith(): self
{
    $this->position = 'begins';

    return $this;
}

public function endsWith(): self
{
     $this->position = 'ends';
}

private function pattern(string $value): string
{
    if ($this->position === 'begins') {
        return $value . '%';
    }

    // etc...
}
```

Not sure about the method names, but hopefully that gives you an idea of the approach I'm suggesting!